server/rooms.ts

```diff
@@ -592,11 +592,8 @@ export abstract class BasicRoom {
 		this.destroy();
 	}
 	reportJoin(type: 'j' | 'l' | 'n', entry: string, user: User) {
-		let reportJoins = this.reportJoins;
-		if (reportJoins && this.settings.modchat && !user.authAtLeast(this.settings.modchat, this)) {
-			reportJoins = false;
-		}
-		if (reportJoins) {
+		const canTalk = user.authAtLeast(this.settings.modchat ?? 'unlocked', this) && !this.isMuted(user);
+		if (this.reportJoins && canTalk) {
```

Comment:
Oh, this code is a lot better.

Does this cover semilocks?

Reply:
Looks like no, can you add that condition? Maybe with an exception for players (i.e. any roomauth)

Reply:
```ts
authAtLeast(minAuth: string, room: BasicRoom | null = null) {
		if (!minAuth || minAuth === ' ') return true;
		if (this.locked || this.semilocked) return false;
...
```
`User#authAtLeast` will return false if a user is semilocked. I can still add an exception for players / voices in the battle, if you would like.

Reply:
Ohhh, yeah, that works.

I do think semilocked players should be able to talk in their own battles, yes.

Reply:
They wouldn't be able to rename spam anyway, because after the first rename they're no longer players.